Utils/src/HiddenSectorProducer.cc

```diff
@@ -22,143 +27,390 @@
+    else{
+      if(isDark(DarkQuarkIDs_,dau)){
+        firstQdM.insert(dau);
+        if(secondDM == false){
+          firstQdM1 = dau;
+          secondDM = true;
+        }
+        else firstQdM2 = dau;
+      }
+      else if(isDark(SMQuarkIDs_,dau)){
+        firstQsM.insert(dau);
+        if(secondSM == false){
+          firstQsM1 = dau;
+          secondSM = true;
+        }
+        else firstQsM2 = dau;
+      }
+    }
+  }
+}
+
+void HiddenSectorProducer::firstDark(CandPtr part, CandSet& firstMd, CandSet& firstQd, CandSet& firstGd, CandSet& firstQdM, CandSet& firstQsM, CandPtr& firstQdM1, CandPtr& firstQdM2, CandPtr& firstQsM1, CandPtr& firstQsM2, bool& secondDM, bool& secondSM) const {
+  if(isDark(DarkFirstIDs_,part)){
+    CandPtr parent = part->mother(0);
+    if(isDark(DarkFirstIDs_,parent)) firstDark(parent, firstMd, firstQd, firstGd, firstQdM, firstQsM, firstQdM1, firstQdM2, firstQsM1, firstQsM2, secondDM, secondSM);
+    else{
+      //SM parent of first dark particles: fill first two vectors here, use mediators to fill second two vectors
+      for(unsigned i = 0; i < part->numberOfDaughters(); i++){
+        CandPtr dau = part->daughter(i);
+        if (isDark(DarkMediatorIDs_,dau)){
+          firstMd.insert(dau);
+          medDecay(dau,firstQdM,firstQsM,firstQdM1,firstQdM2,firstQsM1,firstQsM2,secondDM,secondSM);
+        }
+        else if (isDark(DarkQuarkIDs_,dau)) firstQd.insert(dau);
+        else if (isDark(DarkGluonIDs_,dau)) firstGd.insert(dau);
+      }
+    }
+  }
+}
+
+int HiddenSectorProducer::checkLast(const reco::GenJet& jet, CandSet stableDs, int value, double& frac) const {
+  //compare jet constituents to set of last particles
+  //also compute dark pt fraction in same loop
+  bool match = false;
+  LorentzVector p4;
+  LorentzVector totPt = jet.p4();
+  for(unsigned i = 0; i < jet.numberOfDaughters(); ++i){
+    CandPtr dau = jet.daughter(i);
+    if(isAncestor(DarkHadronIDs_,dau)){
+        match = true;
+        p4 += dau->p4();
+    }
+  }
+  for(const auto& part : stableDs){
+     if(reco::deltaR(jet, *part) < coneSize_)
+     {
+       p4 += part->p4();
+       totPt += part->p4();
+     }
+  }
+  frac = p4.pt()/totPt.pt();
+  return match ? value : 0;
+}
+
+int HiddenSectorProducer::checkFirst(const reco::GenJet& jet, const CandSet& firstP, int value) const {
+  //compare first particles using jet cone (can't compare constituents' parents because all dark hadrons descend from all first dark particles, strong force problems)
+  for(const auto& part : firstP){
+    if(reco::deltaR(jet, *part) < coneSize_) return value;
+  }
+  return 0;
+}
+
+// match particles from mediator to jets
+void HiddenSectorProducer::matchPFMtoJet(CandPtr& part, std::vector<LorentzVector>& matchedJets, const reco::GenJet& jet, int& nPartPerJet, int& t_MT2JetID, const int& mt2ID) const {
+  if(reco::deltaR(jet,*part) < coneSize_ and jet.pt() > 100)
```

Comment:
remind me about the reason for the jet pt cut?

Reply:
In a pair production event, we expect 2 mediators, where each mediator decays to 1 dark quark and 1 SM quark. So we expect 2 dark quarks and 2 SM quarks from the mediators in such an event.

When determining which jets to use to compute the MT2, I decided to impose the following conditions:
1. Each jet can only contain exactly one dark/SM quark from the mediators.
2. Each dark/SM quark from the mediators can only fall into the radius of exactly one jet.

Otherwise, there is no way to tell how the energy from the quark is split among the jets that are sharing the same quark or how to split the energy from the jet if more than one quark falls inside the jet. Therefore events that do not fulfill the conditions above are simply vetoed when it comes to computing the MT2.

If there is no pT requirement for the jet, we see a lot of overlap between jets. This overlap increases the likelihood of violating the conditions, so we end up with less events for computing the MT2.

Reply:
Maybe this is only true for the signal (3000_20_0.3_peak) I have been testing. I guess for a much lower mMed, the pT requirement might become too high for the jet...

Reply:
Yes, I understand what you mean about having multiple jets matching the mediator daughters. How well does it work if you just keep the highest-pT matching jet, ignoring any others? Or, alternatively, just use all the matching jets (adding up their four-vectors)...

Reply:
I guess the most fundamental problem is if daughters from two different mediators are too close together. Maybe those are the cases we should exclude.

Reply:
I think I will remove the pT cut for now.